omniscidb/QueryBuilder/QueryBuilder.cpp

```diff
@@ -0,0 +1,3072 @@
+      {"milli", DateAddField::kMilli},
+      {"millisecond", DateAddField::kMilli},
+      {"milliseconds", DateAddField::kMilli},
+      {"us", DateAddField::kMicro},
+      {"micro", DateAddField::kMicro},
+      {"microsecond", DateAddField::kMicro},
+      {"microseconds", DateAddField::kMicro},
+      {"ns", DateAddField::kNano},
+      {"nano", DateAddField::kNano},
+      {"nanosecond", DateAddField::kNano},
+      {"nanoseconds", DateAddField::kNano},
+      {"week", DateAddField::kWeek},
+      {"weeks", DateAddField::kWeek},
+      {"quarterday", DateAddField::kQuarterDay},
+      {"quarter_day", DateAddField::kQuarterDay},
+      {"quarter day", DateAddField::kQuarterDay},
+      {"quarterdays", DateAddField::kQuarterDay},
+      {"quarter_days", DateAddField::kQuarterDay},
+      {"quarter days", DateAddField::kQuarterDay},
+      {"weekday", DateAddField::kWeekDay},
+      {"week_day", DateAddField::kWeekDay},
+      {"week day", DateAddField::kWeekDay},
+      {"weekdays", DateAddField::kWeekDay},
+      {"week_days", DateAddField::kWeekDay},
+      {"week days", DateAddField::kWeekDay},
+      {"dayofyear", DateAddField::kDayOfYear},
+      {"day_of_year", DateAddField::kDayOfYear},
+      {"day of year", DateAddField::kDayOfYear},
+      {"doy", DateAddField::kDayOfYear}};
+  auto canonical = boost::trim_copy(boost::to_lower_copy(field));
+  if (!field_names.count(canonical)) {
+    throw InvalidQueryError() << "Cannot parse date add field: '" << field << "'";
+  }
+  return field_names.at(canonical);
+}
+
+JoinType parseJoinType(const std::string& join_type) {
+  static const std::unordered_map<std::string, JoinType> join_types = {
+      {"left", JoinType::LEFT},
+      {"inner", JoinType::INNER},
+      {"semi", JoinType::SEMI},
+      {"anti", JoinType::ANTI}};
+  auto canonical = boost::trim_copy(boost::to_lower_copy(join_type));
+  if (!join_types.count(canonical)) {
+    throw InvalidQueryError() << "Cannot parse join type: '" << join_type << "'";
+  }
+  return join_types.at(canonical);
+}
+
+}  // namespace
+
+BuilderExpr::BuilderExpr(const QueryBuilder* builder,
+                         ExprPtr expr,
+                         const std::string& name,
+                         bool auto_name)
+    : builder_(builder), expr_(expr), name_(name), auto_name_(auto_name) {}
+
+BuilderExpr BuilderExpr::rename(const std::string& name) const {
+  return {builder_, expr_, name, false};
+}
+
+BuilderExpr BuilderExpr::avg() const {
+  if (!expr_->type()->isNumber()) {
+    throw InvalidQueryError() << "Unsupported type for avg aggregate: "
+                              << expr_->type()->toString();
+  }
+  auto agg =
+      makeExpr<AggExpr>(builder_->ctx_.fp64(), AggType::kAvg, expr_, false, nullptr);
+  auto name = name_.empty() ? "avg" : name_ + "_avg";
+  return {builder_, agg, name, true};
+}
+
+BuilderExpr BuilderExpr::min() const {
+  if (!expr_->type()->isNumber() && !expr_->type()->isDateTime()) {
```

Comment:
No strings?

Reply:
We don't support min/max aggregate for strings.